federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -0,0 +1,885 @@
+	var hasErr bool
+	for clusterName, cache := range s.clusterCache.clientMap {
+		go func(cache *clusterCache, clusterName string) {
+			err = s.processServiceForCluster(cachedService, clusterName, service, cache.clientset)
+			if err != nil {
+				hasErr = true
+			}
+		}(cache, clusterName)
+	}
+	if hasErr {
+		// detail error has been dumpped inside the loop
+		return fmt.Errorf("service %s/%s was not successfully updated to all clusters", service.Namespace, service.Name), retryable
+	}
+	return nil, !retryable
+}
+
+func (s *ServiceController) deleteFederationService(cachedService *cachedService) (error, bool) {
+	// handle available clusters one by one
+	var hasErr bool
+	for clusterName, cluster := range s.clusterCache.clientMap {
+		err, _ := s.deleteClusterService(clusterName, cachedService, cluster.clientset)
+		if err != nil {
+			hasErr = true
+		}
+	}
+	if hasErr {
+		// detail error has been dumpped inside the loop
+		return fmt.Errorf("service %s/%s was not successfully updated to all clusters", cachedService.lastState.Namespace, cachedService.lastState.Name), retryable
+	}
+	return nil, !retryable
+}
+
+func (s *ServiceController) deleteClusterService(clusterName string, cachedService *cachedService, clientset *clientset.Clientset) (error, bool) {
+	service := cachedService.lastState
+	glog.V(4).Infof("Deleting service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		err = clientset.Core().Services(service.Namespace).Delete(service.Name, &api.DeleteOptions{})
+		if err == nil || errors.IsNotFound(err) {
+			glog.V(4).Infof("service %s/%s deleted from cluster %s", service.Namespace, service.Name, clusterName)
+			return nil, !retryable
+		}
+		time.Sleep(cachedService.nextRetryDelay())
+	}
+	glog.V(4).Infof("failed to delete service %s/%s from cluster %s, %+v", service.Namespace, service.Name, clusterName, err)
+	return err, retryable
+}
+
+func (s *ServiceController) ensureClusterService(cachedService *cachedService, clusterName string, service *api.Service, client *clientset.Clientset) error {
+	var err error
+	var needUpdate bool
+	for i := 0; i < clientRetryCount; i++ {
+		svc, err := client.Core().Services(service.Namespace).Get(service.Name)
+		if err == nil {
+			// service exists
+			glog.V(5).Infof("found service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
+			//reserve immutable fields
+			service.Spec.ClusterIP = svc.Spec.ClusterIP
+
+			//reserve auto assigned field
+			for i, oldPort := range svc.Spec.Ports {
+				for _, port := range service.Spec.Ports {
+					if port.NodePort == 0 {
+						if !portEqualExcludeNodePort(&oldPort, &port) {
+							svc.Spec.Ports[i] = port
+							needUpdate = true
+						}
+					} else {
+						if !portEqualForLB(&oldPort, &port) {
+							svc.Spec.Ports[i] = port
+							needUpdate = true
+						}
+					}
+				}
+			}
+
+			if needUpdate {
+				// we only apply spec update for now
```

Comment:
why "for now". When will we ever update the status?

If we ever update status then there is a chance that federation service controller and k8s service controller can end up overwriting each others status updates so we should prevent that if possible.